### ISSUE TITLE
[BUG] our build system should disallow unused function results

### ISSUE THREAD

**Issue by sergey-shandar** [Azure/azure-sdk-for-c#110]
**Describe the bug**
Our build system (compiler, linter, whatever) should give an error in cases when we or our customers don't use result of a function call.

**To Reproduce**
Remove any `AZ_RETURN_IF_FAILED` in our code base. A compiler will not give you an error or warning.

***Code Snippet***
```c
az_result foo() { return AZ_ERROR; }

az_result bar() {
  foo(); // we expect a build error here
  return AZ_OK;
}
```

**Possible Solutions**

1. use a macro `AZ_NODISCARD` in every function declaration. The macro will use compiler specific pragma/attributes.
    ```c
    #define AZ_NODISCARD special_attribbute_pragma_which_depends_on_compiler

    AZ_NODISCARD az_result foo() { return AZ_ERROR; }
    AZ_NODISCARD az_result bar() {
         foo(); // compilation error
         return AZ_OK;
    } 
    ```
1. use a macro `AZ_RESULT` for every function that return `az_result`. The macro will use compiler specific pragma/attributes.
    ```c
    #define AZ_RESULT special_attribbute_pragma_which_depends_on_compiler az_result

    AZ_RESULT foo() { return AZ_ERROR; }
    AZ_RESULT bar() {
         foo(); // compilation error
         return AZ_OK;
    }     
    ```
1. use some kind of linter tools.

Either option # 1 or # 2 could be a good choice for consumers of Azure SDK for C.

**Comment by barcharcraz** [comment]
I strongly prefer the first option.

the second is VERY confusing and will lead to people using AZ_RESULT as if it was a type, and being surprised when it's not...

I would recommend `AZ_NODISCARD` instead of `AZ_FUNC`

**Comment by barcharcraz** [comment]
the third is strictly worse since we'll have to exclude a bunch of third party stuff from the linter tools anyways, and they won't be able to run on customer code if that customer ignores lots of results from other libraries.

**Comment by barcharcraz** [comment]
fwiw we (the libs team) _very_ liberally sprinkle _NODISCARDs in the stl (this expands to the compiler specific thing or the standard c++ [[nodiscard]])

Also don't use this name, it's reserved.

**Comment by sergey-shandar** [comment]
@barcharcraz Agree. I've renamed `AZ_FUNC` to `AZ_DISCARD`.

**Comment by sergey-shandar** [comment]
@barcharcraz in which order we should apply macros `AZ_INLINE AZ_NODISCARD` or `AZ_NODISCARD AZ_INLINE` ? Or should we have another macro `AZ_INLINE_NODISCARD` ? :-)

**Comment by sergey-shandar** [comment]
Fixed in #113
